File: data/generate_dataset.ts
import fs from 'fs';
import path from 'path';

export interface SyntheticPaymentRecord {
  id: string;
  transaction_id: string;
  merchant_id: string;
  customer_id: string;
  customer_name: string;
  customer_email: string;
  customer_phone: string;
  amount: number;
  currency: string;
  payment_method: 'CARD_CREDIT' | 'CARD_DEBIT' | 'UPI' | 'NETBANKING' | 'WALLET' | 'AUTO_DEBIT';
  payment_status: 'SUCCESSFUL' | 'FAILED' | 'PENDING' | 'RECOVERED' | 'ABANDONED';
  failure_code: string | null;
  failure_reason: string | null;
  failure_category:
    | 'INSUFFICIENT_FUNDS'
    | 'BANK_DOWNTIME'
    | 'EXPIRED_CARD'
    | 'AUTHENTICATION_DROP'
    | 'CUSTOMER_ABANDONMENT'
    | 'FRAUD_SUSPICION'
    | 'GATEWAY_ERROR'
    | 'LIMIT_EXCEEDED'
    | 'NONE';
  customer_age_days: number;
  previous_successful_payments: number;
  previous_failed_payments: number;
  previous_total_spend: number;
  retry_count: number;
  subscription_status: 'ACTIVE' | 'PAST_DUE' | 'UNPAID' | 'CANCELLED' | 'NONE';
  invoice_status: 'PAID' | 'ISSUED' | 'PAST_DUE' | 'UNCOLLECTIBLE' | 'NONE';
  checkout_status: 'COMPLETED' | 'DROPPED' | 'EXPIRED' | 'SESSION_TIMEOUT';
  recovery_status: 'NOT_APPLICABLE' | 'AT_RISK' | 'RECOVERING' | 'RECOVERED' | 'UNRECOVERABLE' | 'EXPIRED';
  recovered_amount: number;
  ml_recovery_probability: number;
  recommended_strategy:
    | 'SMART_RETRY_OFFPEAK'
    | 'DUNNING_WHATSAPP'
    | 'DUNNING_EMAIL'
    | 'PAYMENT_LINK_SMS'
    | 'METHOD_SWITCH_UPI'
    | 'ONE_CLICK_MANDATE_UPDATE'
    | 'MANUAL_INTERVENTION'
    | 'NONE';
  created_at: string;
}

const FIRST_NAMES = ['Aarav', 'Vihaan', 'Aditi', 'Diya', 'Rohan', 'Ananya', 'Kavya', 'Rahul', 'Sneha', 'Vikram', 'Pooja', 'Arjun', 'Isha', 'Siddharth', 'Meera', 'Rishi', 'Neha', 'Kabir', 'Tanvi', 'Aditya'];
const LAST_NAMES = ['Sharma', 'Verma', 'Patel', 'Reddy', 'Gupta', 'Iyer', 'Nair', 'Mehta', 'Joshi', 'Chopra', 'Kapoor', 'Singh', 'Deshmukh', 'Menon', 'Bhat', 'Rao', 'Agarwal', 'Chatterjee', 'Sen', 'Pillai'];
const MERCHANTS = [
  { id: 'mch_saas_cloud', name: 'CloudScale SaaS India', cat: 'B2B Software' },
  { id: 'mch_edtech_pro', name: 'LearnFlow EdTech', cat: 'Education' },
  { id: 'mch_ecom_luxe', name: 'NovaStore Retail', cat: 'E-Commerce' },
  { id: 'mch_fin_stream', name: 'WealthPulse Premium', cat: 'FinTech Subscriptions' }
];

const FAILURE_SCENARIOS = [
  {
    category: 'INSUFFICIENT_FUNDS' as const,
    code: 'BAD_REQUEST_INSUFFICIENT_FUNDS',
    reason: 'Account balance insufficient to complete debit',
    method: 'AUTO_DEBIT' as const,
    recoverability: 0.82,
    strategy: 'SMART_RETRY_OFFPEAK' as const,
    weight: 0.32
  },
  {
    category: 'BANK_DOWNTIME' as const,
    code: 'GATEWAY_ERROR_ISSUER_DOWN',
    reason: 'Issuer bank switch unavailable or timing out',
    method: 'NETBANKING' as const,
    recoverability: 0.89,
    strategy: 'METHOD_SWITCH_UPI' as const,
    weight: 0.22
  },
  {
    category: 'EXPIRED_CARD' as const,
    code: 'CARD_ERROR_EXPIRED',
    reason: 'Card validity date has lapsed or card replaced',
    method: 'CARD_CREDIT' as const,
    recoverability: 0.74,
    strategy: 'ONE_CLICK_MANDATE_UPDATE' as const,
    weight: 0.14
  },
  {
    category: 'AUTHENTICATION_DROP' as const,
    code: 'AUTH_ERROR_OTP_TIMEOUT',
    reason: 'Customer did not enter 3D Secure OTP in time',
    method: 'CARD_DEBIT' as const,
    recoverability: 0.68,
    strategy: 'PAYMENT_LINK_SMS' as const,
    weight: 0.16
  },
  {
    category: 'CUSTOMER_ABANDONMENT' as const,
    code: 'CHECKOUT_USER_DROPOUT',
    reason: 'Customer closed browser checkout iframe before authorization',
    method: 'UPI' as const,
    recoverability: 0.58,
    strategy: 'DUNNING_WHATSAPP' as const,
    weight: 0.10
  },
  {
    category: 'LIMIT_EXCEEDED' as const,
    code: 'CARD_ERROR_TXN_LIMIT',
    reason: 'Daily transaction limit exceeded for card/mandate',
    method: 'CARD_CREDIT' as const,
    recoverability: 0.62,
    strategy: 'DUNNING_EMAIL' as const,
    weight: 0.04
  },
  {
    category: 'FRAUD_SUSPICION' as const,
    code: 'RISK_DECLINE_HIGH_VELOCITY',
    reason: 'Velocity check triggered by issuer fraud defense',
    method: 'CARD_CREDIT' as const,
    recoverability: 0.12,
    strategy: 'MANUAL_INTERVENTION' as const,
    weight: 0.02
  }
];

function randomChoice<T>(arr: T[]): T {
  return arr[Math.floor(Math.random() * arr.length)];
}

function randomWeightedFailure() {
  const rand = Math.random();
  let cumulative = 0;
  for (const s of FAILURE_SCENARIOS) {
    cumulative += s.weight;
    if (rand <= cumulative) return s;
  }
  return FAILURE_SCENARIOS[0];
}

export function generateSyntheticDataset(count = 5000): SyntheticPaymentRecord[] {
  const records: SyntheticPaymentRecord[] = [];
  const now = Date.now();
  const thirtyDaysMs = 30 * 24 * 60 * 60 * 1000;

  // Pre-generate 1,000 distinct customers
  const customers = Array.from({ length: 1200 }, (_, i) => {
    const fn = randomChoice(FIRST_NAMES);
    const ln = randomChoice(LAST_NAMES);
    const id = `cust_${(i + 1000).toString()}`;
    const email = `${fn.toLowerCase()}.${ln.toLowerCase()}${Math.floor(Math.random() * 90 + 10)}@example.com`;
    const phone = `+9198${Math.floor(Math.random() * 90000000 + 10000000)}`;
    const ageDays = Math.floor(Math.random() * 700 + 15);
    const prevSuccess = Math.floor(Math.random() * 25);
    const prevFail = Math.floor(Math.random() * 4);
    const prevSpend = prevSuccess * Math.floor(Math.random() * 4000 + 499);
    return { id, name: `${fn} ${ln}`, email, phone, ageDays, prevSuccess, prevFail, prevSpend };
  });

  for (let i = 0; i < count; i++) {
    const customer = randomChoice(customers);
    const merchant = randomChoice(MERCHANTS);
    const recordId = `pay_${(i + 1).toString().padStart(6, '0')}`;
    const txnId = `txn_rzp_${(10000000 + i).toString()}`;
    
    // Spread timestamp across last 30 days
    const timeOffset = Math.floor(Math.random() * thirtyDaysMs);
    const createdAtIso = new Date(now - timeOffset).toISOString();

    // Realistic amount distribution (e.g. ₹499 to ₹45,000)
    let amount = 999;
    const amtTier = Math.random();
    if (amtTier < 0.40) amount = randomChoice([499, 799, 999, 1499, 1999]);
    else if (amtTier < 0.75) amount = randomChoice([2999, 4999, 7499, 9999]);
    else if (amtTier < 0.95) amount = randomChoice([14999, 19999, 24999, 29999]);
    else amount = Math.floor(Math.random() * 20000 + 30000);

    // Realistic outcome distribution: 68% Initial Success, 32% Failure / Drop
    const isSuccess = Math.random() < 0.68;

    if (isSuccess) {
      records.push({
        id: recordId,
        transaction_id: txnId,
        merchant_id: merchant.id,
        customer_id: customer.id,
        customer_name: customer.name,
        customer_email: customer.email,
        customer_phone: customer.phone,
        amount,
        currency: 'INR',
        payment_method: randomChoice(['UPI', 'CARD_CREDIT', 'CARD_DEBIT', 'NETBANKING', 'AUTO_DEBIT']),
        payment_status: 'SUCCESSFUL',
        failure_code: null,
        failure_reason: null,
        failure_category: 'NONE',
        customer_age_days: customer.ageDays,
        previous_successful_payments: customer.prevSuccess,
        previous_failed_payments: customer.prevFail,
        previous_total_spend: customer.prevSpend,
        retry_count: 0,
        subscription_status: Math.random() > 0.4 ? 'ACTIVE' : 'NONE',
        invoice_status: Math.random() > 0.5 ? 'PAID' : 'NONE',
        checkout_status: 'COMPLETED',
        recovery_status: 'NOT_APPLICABLE',
        recovered_amount: 0,
        ml_recovery_probability: 0.98,
        recommended_strategy: 'NONE',
        created_at: createdAtIso
      });
    } else {
      // Failed or abandoned payment
      const scenario = randomWeightedFailure();
      const retryCount = Math.floor(Math.random() * 3);
      
      // Recovery outcome calculation
      const isRecovered = Math.random() < scenario.recoverability;
      const isRecovering = !isRecovered && Math.random() < 0.55;
      
      let paymentStatus: 'FAILED' | 'RECOVERED' | 'ABANDONED' = 'FAILED';
      let recoveryStatus: 'AT_RISK' | 'RECOVERING' | 'RECOVERED' | 'UNRECOVERABLE' = 'AT_RISK';
      let recoveredAmount = 0;

      if (isRecovered) {
        paymentStatus = 'RECOVERED';
        recoveryStatus = 'RECOVERED';
        recoveredAmount = amount;
      } else if (isRecovering) {
        paymentStatus = 'FAILED';
        recoveryStatus = 'RECOVERING';
      } else {
        paymentStatus = scenario.category === 'CUSTOMER_ABANDONMENT' ? 'ABANDONED' : 'FAILED';
        recoveryStatus = scenario.category === 'FRAUD_SUSPICION' ? 'UNRECOVERABLE' : 'AT_RISK';
      }

      records.push({
        id: recordId,
        transaction_id: txnId,
        merchant_id: merchant.id,
        customer_id: customer.id,
        customer_name: customer.name,
        customer_email: customer.email,
        customer_phone: customer.phone,
        amount,
        currency: 'INR',
        payment_method: scenario.method,
        payment_status: paymentStatus,
        failure_code: scenario.code,
        failure_reason: scenario.reason,
        failure_category: scenario.category,
        customer_age_days: customer.ageDays,
        previous_successful_payments: customer.prevSuccess,
        previous_failed_payments: customer.prevFail + 1,
        previous_total_spend: customer.prevSpend,
        retry_count: retryCount,
        subscription_status: Math.random() > 0.4 ? (isRecovered ? 'ACTIVE' : 'PAST_DUE') : 'NONE',
        invoice_status: Math.random() > 0.5 ? (isRecovered ? 'PAID' : 'PAST_DUE') : 'NONE',
        checkout_status: scenario.category === 'CUSTOMER_ABANDONMENT' ? 'DROPPED' : 'COMPLETED',
        recovery_status: recoveryStatus,
        recovered_amount: recoveredAmount,
        ml_recovery_probability: Number((scenario.recoverability + (Math.random() * 0.1 - 0.05)).toFixed(4)),
        recommended_strategy: scenario.strategy,
        created_at: createdAtIso
      });
    }
  }

  return records;
}

// Generate file if executed directly
if (process.argv[1]?.includes('generate_dataset.ts')) {
  console.log('Generating 5,000 synthetic payments...');
  const dataset = generateSyntheticDataset(5000);
  const outDir = path.join(process.cwd(), 'data');
  if (!fs.existsSync(outDir)) {
    fs.mkdirSync(outDir, { recursive: true });
  }
  const outFile = path.join(outDir, 'synthetic_payments_5000.json');
  fs.writeFileSync(outFile, JSON.stringify(dataset, null, 2), 'utf-8');
  console.log(`Saved ${dataset.length} records to ${outFile}`);
}

File: data/generate_dataset.py
"""
RecoverAI - Synthetic Payment Dataset Generator (Python / Pandas / NumPy)
Generates realistic payment streams, failure categorizations, and recovery signals.
"""

import json
import random
import os
from datetime import datetime, timedelta

MERCHANTS = [
    {"id": "mch_saas_cloud", "name": "CloudScale SaaS India", "category": "B2B Software"},
    {"id": "mch_edtech_pro", "name": "LearnFlow EdTech", "category": "Education"},
    {"id": "mch_ecom_luxe", "name": "NovaStore Retail", "category": "E-Commerce"},
    {"id": "mch_fin_stream", "name": "WealthPulse Premium", "category": "FinTech Subscriptions"},
]

FAILURE_REASONS = [
    {
        "category": "INSUFFICIENT_FUNDS",
        "code": "BAD_REQUEST_INSUFFICIENT_FUNDS",
        "reason": "Account balance insufficient to complete debit",
        "method": "AUTO_DEBIT",
        "recoverability": 0.82,
        "strategy": "SMART_RETRY_OFFPEAK",
        "weight": 0.32,
    },
    {
        "category": "BANK_DOWNTIME",
        "code": "GATEWAY_ERROR_ISSUER_DOWN",
        "reason": "Issuer bank switch unavailable or timing out",
        "method": "NETBANKING",
        "recoverability": 0.89,
        "strategy": "METHOD_SWITCH_UPI",
        "weight": 0.22,
    },
    {
        "category": "EXPIRED_CARD",
        "code": "CARD_ERROR_EXPIRED",
        "reason": "Card validity date has lapsed or card replaced",
        "method": "CARD_CREDIT",
        "recoverability": 0.74,
        "strategy": "ONE_CLICK_MANDATE_UPDATE",
        "weight": 0.14,
    },
    {
        "category": "AUTHENTICATION_DROP",
        "code": "AUTH_ERROR_OTP_TIMEOUT",
        "reason": "Customer did not enter 3D Secure OTP in time",
        "method": "CARD_DEBIT",
        "recoverability": 0.68,
        "strategy": "PAYMENT_LINK_SMS",
        "weight": 0.16,
    },
    {
        "category": "CUSTOMER_ABANDONMENT",
        "code": "CHECKOUT_USER_DROPOUT",
        "reason": "Customer closed browser checkout iframe before authorization",
        "method": "UPI",
        "recoverability": 0.58,
        "strategy": "DUNNING_WHATSAPP",
        "weight": 0.10,
    },
    {
        "category": "LIMIT_EXCEEDED",
        "code": "CARD_ERROR_TXN_LIMIT",
        "reason": "Daily transaction limit exceeded for card/mandate",
        "method": "CARD_CREDIT",
        "recoverability": 0.62,
        "strategy": "DUNNING_EMAIL",
        "weight": 0.04,
    },
    {
        "category": "FRAUD_SUSPICION",
        "code": "RISK_DECLINE_HIGH_VELOCITY",
        "reason": "Velocity check triggered by issuer fraud defense",
        "method": "CARD_CREDIT",
        "recoverability": 0.12,
        "strategy": "MANUAL_INTERVENTION",
        "weight": 0.02,
    },
]


def generate_dataset(num_records=5000):
    records = []
    now = datetime.now()

    # Pre-generate customers
    customers = []
    for i in range(1200):
        c_id = f"cust_{1000 + i}"
        age_days = random.randint(15, 750)
        prev_success = random.randint(0, 30)
        prev_failed = random.randint(0, 5)
        spend = prev_success * random.randint(500, 5000)
        customers.append(
            {
                "id": c_id,
                "name": f"Customer {i+1}",
                "email": f"customer.{i+1}@example.com",
                "phone": f"+9198{random.randint(10000000, 99999999)}",
                "customer_age_days": age_days,
                "previous_successful_payments": prev_success,
                "previous_failed_payments": prev_failed,
                "previous_total_spend": spend,
            }
        )

    for i in range(num_records):
        cust = random.choice(customers)
        mch = random.choice(MERCHANTS)
        txn_id = f"txn_rzp_{10000000 + i}"
        days_ago = random.uniform(0, 30)
        created_at = (now - timedelta(days=days_ago)).isoformat()

        # Amounts
        tier = random.random()
        if tier < 0.4:
            amount = random.choice([499, 799, 999, 1499, 1999])
        elif tier < 0.75:
            amount = random.choice([2999, 4999, 7499, 9999])
        elif tier < 0.95:
            amount = random.choice([14999, 19999, 24999, 29999])
        else:
            amount = random.randint(30000, 50000)

        is_success = random.random() < 0.68

        if is_success:
            record = {
                "id": f"pay_{str(i+1).zfill(6)}",
                "transaction_id": txn_id,
                "merchant_id": mch["id"],
                "customer_id": cust["id"],
                "customer_name": cust["name"],
                "customer_email": cust["email"],
                "customer_phone": cust["phone"],
                "amount": float(amount),
                "currency": "INR",
                "payment_method": random.choice(["UPI", "CARD_CREDIT", "CARD_DEBIT", "NETBANKING", "AUTO_DEBIT"]),
                "payment_status": "SUCCESSFUL",
                "failure_code": None,
                "failure_reason": None,
                "failure_category": "NONE",
                "customer_age_days": cust["customer_age_days"],
                "previous_successful_payments": cust["previous_successful_payments"],
                "previous_failed_payments": cust["previous_failed_payments"],
                "previous_total_spend": float(cust["previous_total_spend"]),
                "retry_count": 0,
                "subscription_status": "ACTIVE" if random.random() > 0.4 else "NONE",
                "invoice_status": "PAID" if random.random() > 0.5 else "NONE",
                "checkout_status": "COMPLETED",
                "recovery_status": "NOT_APPLICABLE",
                "recovered_amount": 0.0,
                "ml_recovery_probability": 0.98,
                "recommended_strategy": "NONE",
                "created_at": created_at,
            }
        else:
            # Weighted failure pick
            weights = [s["weight"] for s in FAILURE_REASONS]
            scenario = random.choices(FAILURE_REASONS, weights=weights, k=1)[0]
            is_recovered = random.random() < scenario["recoverability"]
            is_recovering = (not is_recovered) and (random.random() < 0.55)

            if is_recovered:
                p_status = "RECOVERED"
                r_status = "RECOVERED"
                rec_amount = float(amount)
            elif is_recovering:
                p_status = "FAILED"
                r_status = "RECOVERING"
                rec_amount = 0.0
            else:
                p_status = "ABANDONED" if scenario["category"] == "CUSTOMER_ABANDONMENT" else "FAILED"
                r_status = "UNRECOVERABLE" if scenario["category"] == "FRAUD_SUSPICION" else "AT_RISK"
                rec_amount = 0.0

            record = {
                "id": f"pay_{str(i+1).zfill(6)}",
                "transaction_id": txn_id,
                "merchant_id": mch["id"],
                "customer_id": cust["id"],
                "customer_name": cust["name"],
                "customer_email": cust["email"],
                "customer_phone": cust["phone"],
                "amount": float(amount),
                "currency": "INR",
                "payment_method": scenario["method"],
                "payment_status": p_status,
                "failure_code": scenario["code"],
                "failure_reason": scenario["reason"],
                "failure_category": scenario["category"],
                "customer_age_days": cust["customer_age_days"],
                "previous_successful_payments": cust["previous_successful_payments"],
                "previous_failed_payments": cust["previous_failed_payments"] + 1,
                "previous_total_spend": float(cust["previous_total_spend"]),
                "retry_count": random.randint(0, 3),
                "subscription_status": "ACTIVE" if (is_recovered and random.random() > 0.4) else "PAST_DUE",
                "invoice_status": "PAID" if is_recovered else "PAST_DUE",
                "checkout_status": "DROPPED" if scenario["category"] == "CUSTOMER_ABANDONMENT" else "COMPLETED",
                "recovery_status": r_status,
                "recovered_amount": rec_amount,
                "ml_recovery_probability": round(scenario["recoverability"] + random.uniform(-0.05, 0.05), 4),
                "recommended_strategy": scenario["strategy"],
                "created_at": created_at,
            }

        records.append(record)

    return records


if __name__ == "__main__":
    os.makedirs("data", exist_ok=True)
    dataset = generate_dataset(5000)
    with open("data/synthetic_payments_5000.json", "w", encoding="utf-8") as f:
        json.dump(dataset, f, indent=2)
    print(f"Generated {len(dataset)} records into data/synthetic_payments_5000.json")
